src/platform/backends/hyperv_api/hcn/hyperv_hcn_api_wrapper.cpp

```diff
@@ -0,0 +1,353 @@
+
+namespace multipass::hyperv::hcn
+{
+
+struct GuidParseError : multipass::FormattedExceptionBase<>
+{
+    using FormattedExceptionBase<>::FormattedExceptionBase;
+};
+
+namespace
+{
+
+using UniqueHcnNetwork =
+    std::unique_ptr<std::remove_pointer_t<HCN_NETWORK>, decltype(HCNAPITable::CloseNetwork)>;
+using UniqueHcnEndpoint =
+    std::unique_ptr<std::remove_pointer_t<HCN_ENDPOINT>, decltype(HCNAPITable::CloseEndpoint)>;
+using UniqueCotaskmemString = std::unique_ptr<wchar_t, decltype(HCNAPITable::CoTaskMemFree)>;
+
+namespace mpl = logging;
+using lvl = mpl::Level;
+
+// ---------------------------------------------------------
+
+/**
+ * Category for the log messages.
+ */
+constexpr static auto kLogCategory = "HyperV-HCN-Wrapper";
+
+// ---------------------------------------------------------
+
+/**
+ * Parse given GUID string into a GUID struct.
+ *
+ * @param guid_str GUID in string form, either 36 characters
+ *                  (without braces) or 38 characters (with braces.)
+ *
+ * @return GUID The parsed GUID
+ */
+auto guid_from_string(const std::wstring& guid_wstr) -> ::GUID
+{
+    constexpr static auto kGUIDLength = 36;
+    constexpr static auto kGUIDLengthWithBraces = kGUIDLength + 2;
+
+    const auto input = [&guid_wstr]() {
+        switch (guid_wstr.length())
+        {
+        case kGUIDLength:
+            // CLSIDFromString requires GUIDs to be wrapped with braces.
+            return fmt::format(L"{{{}}}", guid_wstr);
+        case kGUIDLengthWithBraces:
+        {
+            if (guid_wstr.front() != L'{' || guid_wstr.back() != L'}')
+            {
+                throw GuidParseError{"GUID string either does not start or end with a brace."};
+            }
+            return guid_wstr;
+        }
+        }
+        throw GuidParseError{"Invalid length for a GUID string ({}).", guid_wstr.length()};
+    }();
+
+    ::GUID guid = {};
+
+    const auto result = CLSIDFromString(input.c_str(), &guid);
+
+    if (FAILED(result))
+    {
+        throw GuidParseError{"Failed to parse the GUID string ({}).", result};
+    }
+
+    return guid;
+}
+
+/**
+ * Parse given GUID string into a GUID struct.
+ *
+ * @param guid_str GUID in string form, either 36 characters
+ *                  (without braces) or 38 characters (with braces.)
+ *
+ * @return GUID The parsed GUID
+ */
+auto guid_from_string(const std::string& guid_wstr) -> ::GUID
+{
+    const std::wstring v = maybe_widen{guid_wstr};
+    return guid_from_string(v);
+}
+
+// ---------------------------------------------------------
+
+/**
+ * Perform a Host Compute Network API operation
+ *
+ * @param fn The API function pointer
+ * @param args The arguments to the function
+ *
+ * @return HCNOperationResult Result of the performed operation
+ */
+template <typename FnType, typename... Args>
+OperationResult perform_hcn_operation(const HCNAPITable& api, const FnType& fn, Args&&... args)
+{
+    // Ensure that function to call is set.
```

Comment:
This feels redundant. It should be obvious to anyone knowing code. No need to do a live translation between C++ and natural language. That is bound to go out of sync and is distracting to the target audience (us).

Reply:
Removed.